drivers/i2c/i2c_numaker.c

```diff
@@ -0,0 +1,808 @@
+		NUMAKER_PINCTRL_INIT(inst)                                       \
+		.bitrate = DT_INST_PROP(inst, clock_frequency),                  \
+	};                                                                       \
+                                                                                 \
+	static struct i2c_numaker_data i2c_numaker_data_##inst;                  \
+                                                                                 \
+	I2C_DEVICE_DT_INST_DEFINE(inst,                                          \
+			i2c_numaker_init,                                        \
+			NULL,                                                    \
+			&i2c_numaker_data_##inst,                                \
+			&i2c_numaker_config_##inst,                              \
+			POST_KERNEL,                                             \
+			CONFIG_I2C_INIT_PRIORITY,                                \
+			&i2c_numaker_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(I2C_NUMAKER_INIT);
+
+/* Read next data byte, involving ACK/NACK last data byte and message merge */
+static void m_numaker_i2c_master_xfer_msg_read_next_byte(const struct device *dev)
+{
+	const struct i2c_numaker_config *config = dev->config;
+	struct i2c_numaker_data *data = dev->data;
+	I2C_T *i2c_base = config->i2c_base;
+
+	switch (data->master_xfer.buf_end - data->master_xfer.buf_pos) {
+	case 0:
+		/* Last data byte ACKed, we'll do message merge */
+		m_numaker_i2c_master_xfer_msg_end(dev);
+		break;
+	case 1:
+		/* Read last data byte for this message */
+		m_numaker_i2c_master_xfer_msg_read_last_byte(dev);
+		break;
+	default:
+		/* ACK non-last data byte */
+		I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk | I2C_CTL0_AA_Msk);
+	}
+}
+
+/* ACK/NACK last data byte, dependent on whether or not message merge is allowed */
+static void m_numaker_i2c_master_xfer_msg_read_last_byte(const struct device *dev)
+{
+	const struct i2c_numaker_config *config = dev->config;
+	struct i2c_numaker_data *data = dev->data;
+	I2C_T *i2c_base = config->i2c_base;
+
+	/* Shouldn't invoke with message pointer OOB */
+	__ASSERT_NO_MSG(data->master_xfer.msgs_pos < data->master_xfer.msgs_end);
+	/* Should invoke with exactly one data byte remaining for read */
+	__ASSERT_NO_MSG((data->master_xfer.msgs_pos->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ);
+	__ASSERT_NO_MSG((data->master_xfer.buf_end - data->master_xfer.buf_pos) == 1);
+
+	/* Flags of previous message */
+	bool do_stop_prev = (data->master_xfer.msgs_pos->flags & I2C_MSG_STOP);
+
+	/* Advance to next messages temporarily */
+	data->master_xfer.msgs_pos++;
+
+	/* Has next message? */
+	if (data->master_xfer.msgs_pos < data->master_xfer.msgs_end) {
+		/* Flags of next message */
+		struct i2c_msg *msgs_pos = data->master_xfer.msgs_pos;
+		bool is_read_next = ((msgs_pos->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ);
+		bool do_restart_next = (data->master_xfer.msgs_pos->flags & I2C_MSG_RESTART);
```

Comment:
same, no ned for parentehsis on these

Reply:
Thanks of you advice and update for above comments at commit:  01e0dd0d54 & 8e423a12421